pilot/pkg/networking/core/loadbalancer/loadbalancer.go

```
@@ -37,30 +37,24 @@ const (
 func GetLocalityLbSetting(
 	mesh *v1alpha3.LocalityLoadBalancerSetting,
 	destrule *v1alpha3.LocalityLoadBalancerSetting,
-) *v1alpha3.LocalityLoadBalancerSetting {
-	var enabled bool
-	// Locality lb is enabled if its not explicitly disabled in mesh global config
-	if mesh != nil && (mesh.Enabled == nil || mesh.Enabled.Value) {
-		enabled = true
-	}
-	// Unless we explicitly override this in destination rule
+	service *model.Service,
+) (*v1alpha3.LocalityLoadBalancerSetting, bool) {
 	if destrule != nil {
 		if destrule.Enabled != nil && !destrule.Enabled.Value {
-			enabled = false
-		} else {
-			enabled = true
+			return nil, false
 		}
+		return destrule, false
 	}
-	if !enabled {
-		return nil
+	if service != nil && service.Attributes.TrafficDistribution == model.TrafficDistributionPreferClose {
+		return &v1alpha3.LocalityLoadBalancerSetting{
+			Enabled: wrappers.Bool(true),
```

if not other field FailoverPriority set, how do we handle correctly later

---

with just "enabled: true" it enabled fail over automatically. You only need to set `failover` to customize the failover

---

Can you point me where we process it. IMO, in `ApplyLocalityLoadBalancer`, we need other fields. 

```
func ApplyLocalityLoadBalancer(
	loadAssignment *endpoint.ClusterLoadAssignment,
	wrappedLocalityLbEndpoints []*WrappedLocalityLbEndpoints,
	locality *core.Locality,
	proxyLabels map[string]string,
	localityLB *v1alpha3.LocalityLoadBalancerSetting,
	enableFailover bool,
) {
	// before calling this function localityLB.enabled field has been checked.
	if localityLB == nil || loadAssignment == nil {
		return
	}

	// one of Distribute or Failover settings can be applied.
	if localityLB.GetDistribute() != nil {
		applyLocalityWeights(locality, loadAssignment, localityLB.GetDistribute())
		// Failover needs outlier detection, otherwise Envoy will never drop down to a lower priority.
		// Do not apply default failover when locality LB is disabled.
	} else if enableFailover {
		if len(localityLB.FailoverPriority) > 0 {
			// Apply user defined priority failover settings.
			applyFailoverPriorities(loadAssignment, wrappedLocalityLbEndpoints, proxyLabels, localityLB.FailoverPriority)
			// If failover is expliciltly configured with failover priority, apply failover settings also.
			if len(localityLB.Failover) != 0 {
				applyLocalityFailover(locality, loadAssignment, localityLB.Failover)
			}
		} else {
			// Apply default failover settings or user defined region failover settings.
			applyLocalityFailover(locality, loadAssignment, localityLB.Failover)
		}
	}
}
```

---

> apply default failover settings or user defined region failover settings.

this part from the code above.

---

I've added an e2e test as well. But you can see applyLocalityFailover, the `Failover` option is only needed to have 4 priority levels (instead of 3 by default): https://github.com/istio/istio/blob/299c8d1ea8afb9b25534dd19268940f4ede12252/pilot/pkg/networking/core/loadbalancer/loadbalancer.go#L174.

---

LG